pkg/envoy/ads/response.go

```diff
@@ -42,45 +43,50 @@ func (s *Server) sendTypeResponse(tURI envoy.TypeURI,
 	return nil
 }
 
-func (s *Server) sendAllResponses(proxy *envoy.Proxy, server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer, cfg configurator.Configurator) {
-	log.Trace().Msgf("A change announcement triggered *DS update for proxy with SerialNumber=%s on Pod with UID=%s", proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-
-	// Tracks the success of this full update of all its XDS paths. If a single XDS response path fails for this full update,
-	// the full updated will be considered as failed for metric purposes (success = false)
+// sendResponse takes a set of TypeURIs which will be called to generate the xDS resources
+// for, and will have them sent to the proxy server.
+// If no DiscoveryRequest is passed, an empty one for the TypeURI is created
+func (s *Server) sendResponse(typeURIsToSend mapset.Set,
+	proxy *envoy.Proxy,
+	server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer,
+	request *xds_discovery.DiscoveryRequest,
+	cfg configurator.Configurator) error {
 	success := true
-	defer xdsPathTimeTrack(time.Now(), log.Info(), ADSUpdateStr, proxy.GetCertificateSerialNumber().String(), &success)
+	if typeURIsToSend.Cardinality() == len(envoy.XDSResponseOrder) {
+		defer xdsPathTimeTrack(time.Now(), log.Info(), ADSUpdateStr, proxy.GetCertificateSerialNumber().String(), &success)
+	}
 
 	// Order is important: CDS, EDS, LDS, RDS
 	// See: https://github.com/envoyproxy/go-control-plane/issues/59
 	for _, typeURI := range envoy.XDSResponseOrder {
-		// For SDS we need to add ResourceNames
-		var request *xds_discovery.DiscoveryRequest
-		if typeURI == envoy.TypeSDS {
-			request = makeRequestForAllSecrets(proxy, s.catalog)
-			if request == nil {
-				continue
+		if !typeURIsToSend.Contains(typeURI) {
+			continue
+		}
+
+		// Handle request when is not provided, and the SDS case
+		var finalReq *xds_discovery.DiscoveryRequest
+		if request == nil {
```

Comment:
Why would the request be `nil`?

Reply:
Updates forced because of kubernetes changes do not have a discovery request associated from a proxy. If none is passed, it is assumed the request to send data didn't come from a discovery request, however one is generated (this doesn't change from current implementation, see [here](https://github.com/openservicemesh/osm/pull/2844/files#diff-7a1f450fbabb0088828083ed90ad3ccc49ed4edd9ae6f58ad5bdcdd158b2dda8L64))

The main difference is that this function now can take a discovery request coming from a proxy if present, and use that one instead.